build_matrix_A.py

```diff
@@ -0,0 +1,86 @@
+    outputs. List of tuples of format (material, number of nodes)
+    --------
+    - An array that can be plugged into a subsequent function to solve the
+    A.x=b formula
+    """
+    k = np.zeros(len(mesh))
+    rho = np.zeros(len(mesh))
+    c = np.zeros(len(mesh))
+    DR = np.zeros(len(mesh))
+    for m in range(len(mesh)):
+        k[m] += material_property_library[mesh[m][0]]['k']
+        rho[m] += material_property_library[mesh[m][0]]['rho']
+        c[m] += material_property_library[mesh[m][0]]['c']
+        DR[m] += mesh[m][1][1]-mesh[m][1][0]
+
+
+    total_nodes = len(mesh[0][1])
+    for i in range(1, len(mesh)):
+        total_nodes += len(mesh[i][1])-1
+        
+    A = np.zeros((total_nodes, total_nodes))
+    A_test = np.zeros((3,total_nodes))
+    volume = np.zeros(total_nodes)
+    
+    "Begin by setting boundary nodes"
+    "Have to treat nodes 0 and 1 differently to avoid divide by zero"
+    volume[0] = 4/3*np.pi*(mesh[0][1][0]+DR[0]/2)**3
+    volume[1] = 4/3*np.pi*(mesh[0][1][1]+DR[0]/2)**3-(mesh[0][1][1]-DR[0]/2)**3
+    
+    A[0, 0] = -k[0]*4*np.pi/(DR[0]*volume[0])*(DR[0]/2)**2
+    A[0, 1] = k[0]*4*np.pi/(DR[0]*volume[0])*(DR[0]/2)**2
+	
+
+    A[1, 1] = -k[0]*4*np.pi/volume[1]*(1/(1/mesh[0][1][1]-1/mesh[0][1][2]) +
+                                    1/DR[0]*(mesh[0][1][0]+DR[0]/2)**2)
+    A[1, 0] = k[0]*4*np.pi/(DR[0]*volume[1])*(mesh[0][1][0]+DR[0]/2)**2
+    A[1, 2] = k[0]*4*np.pi/volume[1]*(1/(1/mesh[0][1][1]-1/mesh[0][1][2]))
+	
+
+	
+    for i in range(2, len(mesh[0][1])-1): #Length set to fueled node length
+        volume[i] = 4/3*np.pi*((mesh[0][1][i]+DR[0]/2)**3-(mesh[0][1][i]-DR[0]/2)**3)
+        A[i, i] = -k[0]*4*np.pi/volume[i]*(1/(1/mesh[0][1][i]-1/mesh[0][1][i+1]) +
+                                        1/(1/mesh[0][1][i-1]-1/mesh[0][1][i]))
+        A[i, i-1] = k[0]*4*np.pi/volume[i]*(1/(1/mesh[0][1][i-1]-1/mesh[0][1][i]))
+        A[i, i+1] = k[0]*4*np.pi/volume[i]*(1/(1/mesh[0][1][i]-1/mesh[0][1][i+1]))
+		
+    #Now need to handle the boundary nodes
+    #i is currently the value of the node at the internal boundary
+    for mat in range(1, len(mesh)):
+        prev_mesh_length = len(mesh[mat-1][1])
+        volume[i+1] = 4/3*np.pi*((mesh[mat-1][1][prev_mesh_length-1]+DR[mat]/2)**3-(
+                                                 mesh[mat-1][1][prev_mesh_length-1]-DR[mat-1]/2)**3)
+        A[i+1,i+2] = k[mat]*4*np.pi/volume[i+1]*(1/(1/mesh[mat][1][0]-1/mesh[mat][1][1]))
+        A[i+1,i+1] = -k[mat-1]*4*np.pi/volume[i+1]*(1/(1/mesh[mat-1][1][prev_mesh_length-2]
+                                                    -1/mesh[mat-1][1][prev_mesh_length-1]))-(
+                                                    k[mat]*4*np.pi/volume[i+1]*(1/(1/mesh[mat][1][0]
+                                                    -1/mesh[mat][1][1])))
+        A[i+1, i] = k[mat-1]*4*np.pi/volume[i+1]*(1/(1/mesh[mat-1][1][prev_mesh_length-2]-1/mesh[mat-1][1][prev_mesh_length-1]))
+	
```

Comment:
If possible I would like to see the derivation of the difference equations that you are using. It appears that you are using a set of flux balance equations rather than a finite difference approximation of the PDE. I believe that your approach works, but it is hard for me to follow in code. The finite difference approach would likely not have the same algebraic cancellation that you leave in, i.e. 4*pi/4*pi. We could discuss that implementation and check that it works if you are interested. Also, a different choice in mesh structure could provide a uniform way to create the matrix for all interior points. My thought is a dictionary of numpy arrays for 'R', 'DR', 'material'. There are also certainly also disadvantages to this storage model compared to the current one.